desc/objectives/_free_boundary.py

```diff
@@ -0,0 +1,840 @@
+        constants : dict
+            Dictionary of constant data, eg transforms, profiles etc. Defaults to
+            self.constants
+
+        Returns
+        -------
+        f : ndarray
+            Boundary force error (N).
+
+        """
+        if constants is None:
+            constants = self.constants
+        src_data = compute_fun(
+            "desc.equilibrium.equilibrium.Equilibrium",
+            self._eq_data_keys,
+            params=eq_params,
+            transforms=constants["src_transforms"],
+            profiles=constants["src_profiles"],
+        )
+        eval_data = compute_fun(
+            "desc.equilibrium.equilibrium.Equilibrium",
+            self._eq_data_keys,
+            params=eq_params,
+            transforms=constants["eval_transforms"],
+            profiles=constants["eval_profiles"],
+        )
+        if self._sheet_current:
+            assert sheet_params is not None
+            # enforce that they're the same surface
+            sheet_params["R_lmn"] = eq_params["Rb_lmn"]
+            sheet_params["Z_lmn"] = eq_params["Zb_lmn"]
+            sheet_src_data = compute_fun(
+                "desc.magnetic_fields.FourierCurrentPotentialField",
+                self._sheet_data_keys,
+                params=sheet_params,
+                transforms=constants["sheet_src_transforms"],
+                profiles={},
+            )
+            sheet_eval_data = compute_fun(
+                "desc.magnetic_fields.FourierCurrentPotentialField",
+                self._sheet_data_keys,
+                params=sheet_params,
+                transforms=constants["sheet_eval_transforms"],
+                profiles={},
+            )
+            src_data["K_vc"] += sheet_src_data["K"]
+
+        # this is in cartesian
+        Bplasma = -virtual_casing_biot_savart(
+            eval_data,
+            constants["eval_transforms"]["grid"],
+            src_data,
+            constants["src_transforms"]["grid"],
+            constants["interpolator"],
+            loop=self._loop,
+        )
+        # need extra factor of B/2 bc we're evaluating on plasma surface
+        Bplasma = xyz2rpz_vec(Bplasma, phi=eval_data["zeta"]) + eval_data["B"] / 2
+        x = jnp.array([eval_data["R"], eval_data["zeta"], eval_data["Z"]]).T
+        Bext = constants["ext_field"].compute_magnetic_field(
+            x, grid=self._field_grid, basis="rpz"
+        )
+        Bex_total = Bext + Bplasma
+        Bin_total = eval_data["B"]
+        Bn = jnp.sum(Bex_total * eval_data["n_rho"], axis=-1)
+
+        bsq_out = jnp.sum(Bex_total * Bex_total, axis=-1)
+        bsq_in = jnp.sum(Bin_total * Bin_total, axis=-1)
+
+        g = eval_data["|e_theta x e_zeta|"]
+        Bn_err = Bn * g
+        Bsq_err = (bsq_in + eval_data["p"] * (2 * mu_0) - bsq_out) * g
+        Bjump = Bex_total - Bin_total
+        if self._sheet_current:
+            Kerr = mu_0 * sheet_eval_data["K"] - jnp.cross(eval_data["n_rho"], Bjump)
```

Comment:
this Kerr just ensures that the sheet current is the one due to n x B_jump right?

Reply:
yep